_posts/2015-09-12-developing-gamepad-virtual-keyboard.markdown

```diff
@@ -0,0 +1,150 @@
+### The keyboard layout
+
+This is the part I’ll talk about the least, because implementation varies greatly with the type of layout. The layout could be a radial keyboard, like the Daisywheel, a traditional keyboard, like QWERTY, or something completely different like [Beyond Good and Evil’s keyboard](https://youtu.be/euquOpUmUyk)). No matter what layout you have, let’s assume the user can only type one character at a time. This could be accessed by some sort of event or callback that sends the symbol to any part of the app that’s interested. One of those parts is the Input field.
+
+### The inputs
+
+When working with a virtual keyboard, the user has to know what text they’ve currently typed and where their cursor is located in that string of text. For this we’re going to use an HTML input tag. In most cases of a virtual keyboard, there is the input field that is actually part of the form or page, and another input that floats above the virtual keyboard. This second input is used to temporarily show the current text being entered. For the rest of the article I’ll be calling the real input, the *Real Input*, and this temporary one, the *Fake Input*. OnFocus of the Real Input, the keyboard and the Fake Input will be presented. There’s also some other baseline functionality that should be expected from this Fake Input.
+
+ + Shows the cursor's location.
+ + Shows current text, this includes text that was already in the Real Input onFocus.
+ + Places new text directly after the cursor location and the cursor shifts one character to the right.
+ + Cursor can change position using the **mouse**, **keyboard**, OR the **gamepad**.
+ + Text can be entered with the **gamepad** OR **keyboard**.
+
+That first one is easy. The HTML input tag automatically shows the current location of the cursor.
+
+<p style=”text-align: center;”>
+<input placeholder=”this is an input”>
+<p>
+
+The 2nd one is almost as straight-forward. We’ll have to keep track of that current text string value. If you’re using Flux, you’ll probably want to create a Store for the value of the input tag at this point. That way our Fake Input and the Real Input will both mirror the same values when the virtual keyboard is dismissed.
+
+The functionality for those last 3 points all fold into each other, so we’ll discuss them all at once.
+
+Let’s assume you already have a mechanism for getting the currently selected symbol. We'll need a callback for that event, let's call it onGamepadEvent. Inside this callback we receive the new character being typed out. We can't just concatenate it onto the end of our current string because the cursor could be anywhere. So we need to get the current string value from the input, get the cursor location within that string, and then insert the new character at that location of the cursor. We also need to increment the location of the cursor, as I'll discuss next.
+
+```js
+// stores/input.js
+
+getValueDivision: function() {
+    var value = this.value + '';
```

Comment:
why make it an agrument? this helps encapsulate the logic for getting the value parts. Why should `onGamepadEvent` be sending `this.value` along?